### PR TITLE
Cached CompletedTask

### PR DESCRIPTION
**Pull request by Maxwe11**
excluded from #908

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SpecializedTasks.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿namespace StyleCop.Analyzers.Helpers
+{
+    using System.Threading.Tasks;
+
+    internal static class SpecializedTasks
+    {
+        internal static Task CompletedTask { get; } = Task.FromResult(default(VoidResult));
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SpecializedTasks.cs]: The big downside of this compared to `Task.CompletedTask` is we don't create the task with `InternalTaskOptions.DoNotDispose`. If someone calls `SpecializedTasks.CompletedTask.Dispose()`, this property will become unusable until the process is restarted.

Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SpecializedTasks.cs]: `Task.CompletedTask` not supported in portable subset.
We can add one more rule for internal analyzer #813 :smile:

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SpecializedTasks.cs]: > `Task.CompletedTask` not supported in portable subset.

True, but also doesn't address the concern. Has the Roslyn team indicated that they don't dispose of `Task` instances by design?

Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SpecializedTasks.cs]: Good catch. Didn't find any documentation about that except source code:
https://github.com/dotnet/roslyn/blob/918dd32ccdfd901bfd71d88bdb9f7051adfb3d51/src/Features/Core/CodeFixes/CodeFixService.cs#L219

https://github.com/dotnet/roslyn/blob/918dd32ccdfd901bfd71d88bdb9f7051adfb3d51/src/Features/Core/CodeFixes/CodeFixService.cs#L393

https://github.com/dotnet/roslyn/blob/3faf03a3e52c1b46bbc3967060cf6406d4ecc9af/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/BatchFixAllProvider.cs#L111

in all case they use own cached task if user's implementation of `RegisterCodeFixesAsync` returns `null`.

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SpecializedTasks.cs]: :thought_balloon: That's a really interesting find.

**Comment by sharwell** [comment]
This is blocked on dotnet/roslyn#3218.

**Comment by sharwell** [comment]
Merging this based on the comments on the issue linked above along with a review of the implementation. The paths taken for normal async/await operations do not throw if the `Task` is both completed and disposed.

**Pull request merged**